src/vmm/src/builder.rs

```diff
@@ -763,39 +767,543 @@ pub mod tests {
 
-    #[test]
-    fn test_stdin_wrapper() {
-        let wrapper = SerialStdin(io::stdin());
-        assert_eq!(wrapper.as_raw_fd(), io::stdin().as_raw_fd())
+    struct CustomBlockConfig {
+        drive_id: String,
+        is_root_device: bool,
+        partuuid: Option<String>,
+        is_read_only: bool,
     }
 
-    #[test]
-    fn test_setup_serial_device() {
-        // Wrapper over TempFile that implements `ReadableFd`.
-        struct SerialInput(File);
-        impl io::Read for SerialInput {
-            fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-                self.0.read(buf)
+    impl CustomBlockConfig {
+        fn new(drive_id: String, is_root_device: bool, partuuid: Option<String>, is_read_only: bool) -> Self {
+            CustomBlockConfig {
+                drive_id,
+                is_root_device,
+                partuuid,
+                is_read_only,
             }
         }
-        impl AsRawFd for SerialInput {
-            fn as_raw_fd(&self) -> RawFd {
-                self.0.as_raw_fd()
-            }
+    }
+
+    fn default_mmio_device_manager() -> MMIODeviceManager {
+        MMIODeviceManager::new(
+            &mut (arch::MMIO_MEM_START as u64),
+            (arch::IRQ_BASE, arch::IRQ_MAX),
+        )
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn default_portio_device_manager() -> PortIODeviceManager {
+        PortIODeviceManager::new(
+            Arc::new(Mutex::new(Serial::new_sink(
+                EventFd::new(libc::EFD_NONBLOCK).unwrap(),
+            ))),
+            EventFd::new(libc::EFD_NONBLOCK).unwrap(),
+        )
+        .unwrap()
+    }
+
+    fn default_kernel_cmdline() -> Cmdline {
+        let mut kernel_cmdline = kernel::cmdline::Cmdline::new(4096);
+        kernel_cmdline.insert_str(DEFAULT_KERNEL_CMDLINE).unwrap();
+        kernel_cmdline
+    }
+
+    fn default_vmm() -> Vmm {
+        let guest_memory = create_guest_memory(128).unwrap();
+        let kernel_cmdline = default_kernel_cmdline();
+
+        let exit_evt = EventFd::new(libc::EFD_NONBLOCK)
+            .map_err(Error::EventFd)
+            .map_err(StartMicrovmError::Internal)
+            .unwrap();
+
+        let vm = setup_kvm_vm(&guest_memory).unwrap();
+        let mmio_device_manager = default_mmio_device_manager();
+        #[cfg(target_arch = "x86_64")]
+        let pio_device_manager = default_portio_device_manager();
+
+        Vmm {
+            events_observer: Some(Box::new(SerialStdin::get())),
+            guest_memory,
+            kernel_cmdline,
+            vcpus_handles: Vec::new(),
+            exit_evt,
+            vm,
+            mmio_device_manager,
+            #[cfg(target_arch = "x86_64")]
+            pio_device_manager,
+            seccomp_level: SECCOMP_LEVEL_ADVANCED,
+        }
+    }
+
+    fn helper_test_attach_block_device(event_manager: &mut EventManager, custom_block_cfgs: Vec<CustomBlockConfig>) -> (Vmm, Vec<TempFile>) {
```

Comment:
It looks like we never use this helper function for the `block_files` member of the returned tuple. Could we return only the vmm and maybe rename this function to something like `vmm_with_block_devices` or something like this :-??.
Also, why do we need to create `block_files` vec in this function?

Reply:
I will rename the function with your suggestion. The `Vec` is created because `TempFile`s are created inside the for loop scope. Once we get out of that scope, they are destroyed, making `attach_block_devices` return error. Keeping them inside a `Vec`, will make them available for the `attach_block_devices` call, and returning them from `helper..` will pass them to the outer scope of the function call, besides the vmm object. We can dispose them inside `helper_`, and not return them at all, but that will make the vm block devices point to deleted files. This is not a big deal, in the context of these unit tests, since we are not testing block functionality.